src/agent/codecoveragesummaryreader.ts

```diff
@@ -0,0 +1,219 @@
+import ifm = require('./interfaces');
+import testifm = require('vso-node-api/interfaces/TestInterfaces');
+import ccp = require('./codecoveragepublisher');
+import utilities = require('./utilities');
+import cm = require('./common');
+import Q = require('q');
+
+var xmlreader = require('xmlreader');
+
+export class CodeCoverageSummary {
+    results: testifm.CodeCoverageStatistics[];
+
+    constructor() {
+        this.results = [];
+    }
+
+    addResults(res) {
+        this.results = this.results.concat(res);
+    }
+}
+
+export class JacocoSummaryReader implements ccp.ICodeCoverageReader {
+
+    private command: cm.ITaskCommand;
+
+    constructor(command: cm.ITaskCommand) {
+        this.command = command;
+    }
+
+    public getCodeCoverageSummary(summaryFilePath: string): Q.Promise<testifm.CodeCoverageData> {
+        var defer = Q.defer<testifm.CodeCoverageData>();
+        var _this = this;
+
+        SummaryReaderUtilities.getXmlContent(summaryFilePath).then(function(xmlContents) {
+            _this.parseJacocoXmlReport(xmlContents).then(function(codeCoverageSummary) {
+                defer.resolve(SummaryReaderUtilities.getCodeCoverageData(codeCoverageSummary));
+            }).fail(function(err) {
+                defer.reject(err);
+            });
+        }).fail(function(err) {
+            defer.reject(err);
+        });
+
+        return defer.promise;
+    }
+
+    private parseJacocoXmlReport(xmlContent): Q.Promise<CodeCoverageSummary> {
+
+        this.command.info("parseJacocoXmlReport: Parsing summary file.");
+        var defer = Q.defer<CodeCoverageSummary>();
+        if (!xmlContent.report || !xmlContent.report.at(0) || !xmlContent.report.at(0).counter) {
+            defer.resolve(null);
+        }
+        try {
+            var coverage = new CodeCoverageSummary();
+            var reportNode = xmlContent.report.at(0);
+
+            var nodeLength = reportNode.counter.count();
+            var coverageStats = [];
+            for (var i = 0; i < nodeLength; i++) {
+                var counterNode = reportNode.counter.at(i);
+                var attributes = counterNode.attributes();
+                if (attributes && attributes.type && attributes.covered && attributes.missed) {
+                    var total = Number(attributes.covered) + Number(attributes.missed);
+                    this.command.info(attributes.type + " : " + attributes.covered + "/" + total + " covered.");
+                    var coverageStat = SummaryReaderUtilities.getCodeCoverageStatistics(attributes.type, attributes.covered, total, attributes.type);
+                    coverageStats.push(coverageStat);
+                }
+            }
+            coverage.addResults(coverageStats);
+        }
+        catch (error) {
+            defer.reject(error);
+            return defer.promise;
+        }
+
+        defer.resolve(coverage);
```

Comment:
check the if condition above try block. you are resolving to null and here trying to set invalid var. This will throw JS error